Docs for SOCKS5 errors

felipedau opened the pull request:
I believe that mentioning these errors in the SOCKS5 section of the "Programming Guide" would be a good way to present them. I added the `socks` module to the docs containing only these errors because I needed a place for those classes and that looked like the right one to me. I'm not sure which other functions/classes we might want to add to this file because I think that module is a bit abstracted from the user and they might not touch those names. Is that right?

I did not add more details to the errors' docstrings but I will once I have a nice reference for that. For now I wrote that vague note until we have more info about the reasons for each error and tell the user they can find them in that module. (I should be able to do that during the week, so I think we can do everything with this same PR)

What do you think?

meejah commented on docs/guide.rst:
```diff
@@ -391,6 +391,10 @@ There are also lower-level APIs to create
+handled by catching the :class:`txtorcon.socks.SocksError` class. If
+you need to work with each specific error, see the ":ref:`socks`" for
+a list of them.
 
```
You could maybe say something about these corresponding to the RFC-specified SOCKS5 errors.

meejah commented:
> I'm not sure which other functions/classes we might want to add to this file because I think that module is a bit abstracted from the user and they might not touch those names. Is that right?

To answer this directly: anything *without* a leading underscore is good to document. Even the ones *with* an underscore are fine but so the `resolve`, `resolve_ptr` and `TorSocksEndpoint` are things that a user could use. Stuff like `_TorSocksProtocol` etc shouldn't be touched by users-of-txtorcon so I'd leave these out of the API docs for now.

felipedau commented:
> For context, the "SOCKS stuff" should ideally be in its own project really.

It would be great to have it in Twisted. Is anyone planning to build one based on the SOCKS4 client? 

> The "tor specific opcodes" stuff complicates this a little (and there are probably more coming it sounds like).

Yeah, I think in the end txtorcon will have it's own (Tor Extended) SOCKS5 client, either modifying the current one or subclassing from the "ideal" project.

> But yes adding docs for "socks" module is great. Like with other things, prefacing it with "this is low-level stuff, you own both pieces .. " with pointers to the higher-level stuff would be a good idea and then documenting everything in there is fine.

Good idea!

> Basically with the "new docs" I'm hoping that "almost nobody" has to actually delve into the "API Docs" except perhaps a couple clicks to see a function signature etc -- so your paragraph in "guide" looks good.

I see. BTW this new version is great - it's so easy to use. Great job!

> To answer this directly: anything without a leading underscore is good to document. Even the ones with an underscore are fine but so the resolve, resolve_ptr and TorSocksEndpoint are things that a user could use. Stuff like _TorSocksProtocol etc shouldn't be touched by users-of-txtorcon so I'd leave these out of the API docs for now.

Alright, I'll add those as well.

> It sounds like there might be some better error-reporting coming from Tor so that we can tell (via SOCKS) why a hidden-sevice thing failed instead of just getting TtlExpiredError or whatever. See 229-further-socks5-extensions.txt in the torspec repo.
> 
> Anyway I'm happy to merge as-is, or with some additions if you want; let me know :)

Ideally I wanted to have more information about the causes of such errors, but I am not sure if that's possible at the moment. Earlier this month I had a hard time to find a good docs about that even for "regular" SOCKS servers. What I have not understood is whether the errors I get from Tor have the same meaning I would get when using SOCKS without it. For example: for Tor, is a  "general SOCKS server failure" the same thing as it is for a "regular" SOCKS server?

Apparently to document these errors properly, we either need help from someone who worked on the SOCKS5 code in Tor or looking into that ourselves. I think before adding such details to the errors, we should wait until 229 is implemented and talk to someone who can provide better info on this. What do you think?

felipedau commented:
- Is it okay to add a "low-level" section like that?
- Should the language be formal?
- I think it is nice to have the docstrings pointing to the high-level APIs, so I added one to `resolve_ptr`
- Should `resolve_ptr`'s param be `ip` instead of `hostname`? (Trying to make it look just like `dns_resolve_ptr`)

Let me know what you think!

meejah commented:
for the "low level" you could do it in a `.. note:` or similar. I wouldn't say "could break at any time" (because that's only true for underscore-prefixed things) but instead something about preferring the `Tor.*` calls instead.

meejah commented:
I like the pointers to the "high level" docs, and the RFC links!

meejah commented:
For the arg: yes you're right one of them should be `hostname` and the other `ip` or similar.

Thanks again for your work on this!

felipedau commented:
> for the "low level" you could do it in a .. note: or similar. I wouldn't say "could break at any time" (because that's only true for underscore-prefixed things) but instead something about preferring the Tor.* calls instead.

Hmm right, that warning was wrong. Fixed that and made it a bit simpler: I thought the "you know what you are doing" was maybe too informal and unnecessary, so I just left the suggestion of using high-level APIs.

> For the arg: yes you're right one of them should be hostname and the other ip or similar.

Okay, renamed it.

> Thanks again for your work on this!

I'm happy to help :)

I think that for now the information we have is fine. Once we are able to answer the questions from the comment I posted a couple days ago then we can go back to this. I guess. Let me know what you think.

Thanks!

meejah commented:
This looks good to merge to me as-is. If you want to wait for more explanations of the errors, that's fine but that can just be a follow-on PR. Shall I merge?

(Also yes we'll have to just look in Tor source to know what errors correspond to what situations .. and/or make someone else do this for us ;)

felipedau commented:
> This looks good to merge to me as-is. If you want to wait for more explanations of the errors, that's fine but that can just be a follow-on PR. Shall I merge?
> 
> (Also yes we'll have to just look in Tor source to know what errors correspond to what situations .. and/or make someone else do this for us ;)

Absolutely, I think this is ready. Once we have more info about the errors, then we can go back to it.

Thanks!
